ports/qemu-riscv/README.md

```diff
@@ -0,0 +1,73 @@
+This is an experimental, community-supported port for RISC-V RV32IMC emulation
+as provided by QEMU (http://qemu.org).
+
+The purposes of this port are to enable:
+
+1. Continuous integration
+    - run tests against architecture-specific parts of code base
+2. Experimentation
+    - simulation & prototyping of anything that has architecture-specific
+      code
+    - exploring instruction set in terms of optimising some part of
+      MicroPython or a module
+3. Streamlined debugging
+    - no need for JTAG or even an MCU chip itself
+    - no need to use OpenOCD or anything else that might slow down the
+      process in terms of plugging things together, pressing buttons, etc.
+
+This port requires a newlib based bare metal/ELF RISC-V toolchain, either
+with multilib support or 32 bits specific (M, C, and Zicsr extensions must
+be supported, along with ilp32 ABI).
+
+If your toolchain does not support the above requirements, either download
+a pre-built toolchain (and that's a hit or miss situation in its own right)
```

Comment:
Ubuntu 22.04 uses GCC 10 by default when installing `gcc-riscv64-unknown-elf` - the changes you suggest don't work with that :(

```bash
root@mpy-ci:~/micropython/ports/qemu-riscv# CFLAGS="-I/usr/lib/picolibc/riscv64-unknown-elf/include"   make -f Makefile.test test PICOLIBC=1 -j 4
Use make V=1 or set BUILD_VERBOSE in your environment to increase build verbosity.
(cd ../../tests; ./run-tests.py --target=qemu-riscv --write-exp)
# ...
CC ../../lib/libm/roundf.c
CC test_main.c
CC ../../lib/tinytest/tinytest.c
/usr/lib/riscv64-unknown-elf/bin/ld: cannot find crt0.o: No such file or directory
/usr/lib/riscv64-unknown-elf/bin/ld: cannot find EL: No such file or directory
/usr/lib/riscv64-unknown-elf/bin/ld: cannot find -lc
/usr/lib/riscv64-unknown-elf/bin/ld: cannot find -lgloss
collect2: error: ld returned 1 exit status
make: *** [Makefile.test:19: build/firmware-test.elf] Error 1
```

Should I look into adapting your changes to cater for GCC 10?  This is dependent on the minimum compiler version decision.  In the meantime can you tell me which tests fail on your environment?

Reply:
Hmm, I would hope there is a way to make this work via multilib although maybe not given the amount of flux in riscv toolchain land around this time. If you have time to dig into it then please do, I can take a look as well if it's helpful, it'd be pretty good if the Ubuntu 22.04 toolchain could "just work".

Having to manually tell the toolchain about the sysroot include directory seems like it shouldn't be necessary as well, there should be some way to give this information to the toolchain in a high level way - which will then pass through to finding the correct system libraries. In theory, at least. :upside_down_face:

Reply:
> In the meantime can you tell me which tests fail on your environment?

Oops, forgot to answer this question.

```
❯ git describe
v1.22.0-277-gd9f6027f69  # branch on my fork with the two linked commits

❯ make clean
[...]

❯ make -f Makefile.test CROSS_COMPILE=riscv64-elf- test
[...]

❯ grep -a FAILED build/console.out
  [micropython/viper_unop.py FAILED]
  [extmod/re_namedclass.py FAILED]
  [basics/generator_throw.py FAILED]
  [basics/gen_yield_from_throw.py FAILED]
  [basics/slice_op.py FAILED]
  [basics/string1.py FAILED]
  [basics/bytes.py FAILED]
  [float/float2int_intbig.py FAILED]
  [float/float2int_fp30_intbig.py FAILED]
9/616 TESTS FAILED. (175 skipped)
```

Full file: 
[console.out.txt](https://github.com/micropython/micropython/files/15242173/console.out.txt)

(It was 6 tests but then I realised it was comparing output with an old build of micropython unix port, and unfortunately it become 9 failures!)

The failing tests are the same for the two precompiled toolchains mentioned above. I can't tell if the output is 100% the same because the tests ran in a different order each time.[*]

If you don't see these failures with your toolchain then I'm also happy to dig around some more and look for whatever the difference is.

[*] I feel like there's a lot of low hanging fruit for improving the test tooling for qemu. Going to try and look into this...

Reply:
Oh snap, I get exactly the same set of failed tests for the qemu-arm branch on my system. So probably this is something with my qemu, or my system, and not this PR. Apologies!

Reply:
>  In the meantime can you tell me which tests fail on your environment?

Sorry, my mistake, the test failures I saw were mostly PEBKAC (exp files had been generated with CPython 3.12 where a bunch of things fail everywhere, had to roll back to 3.10.)

I still see one failure:

```
micropython/viper_unop.py: 
Traceback (most recent call last):
  File "<stdin>", line 4, in <module>
SyntaxError: invalid micropython decorator

  FAIL ../../shared/upytesthelper/upytesthelper.c:128: Uncaught exception
result: 
  [micropython/viper_unop.py FAILED]
```

This is a pretty new native emitter feature (3c445f663), so guessing it broke after rebase and you'll see the same failure. Let me know if you want me to dig into it more.

Reply:
Ok, viper is not yet supported (I've already written a native emitter for RV32, and this PR is part of that work) :)

Anyway, regarding Ubuntu 22.04, a *partial* fix for the build once the `FLT_EVAL_METHOD` patch lands (and `__PICOLIBC__` is used) could be this:

```diff
diff --git a/ports/qemu-riscv/Makefile b/ports/qemu-riscv/Makefile
index 8e254a053..f2998375a 100644
--- a/ports/qemu-riscv/Makefile
+++ b/ports/qemu-riscv/Makefile
@@ -22,6 +22,17 @@ GCC_MAJOR = $(word 1, $(GCC_VERSION))
 GCC_MINOR = $(word 2, $(GCC_VERSION))
 GCC_PATCH = $(word 3, $(GCC_VERSION))
 
+# Ubuntu 22.04 ships with Picolibc rather than newlib.  To avoid having
+# per-distro workarounds, and given that Picolibc may get more support
+# in the future, it is chosen over newlib if it is found.
+
+PICOLIBC_SPECS = $(shell $(CC) --print-file-name=picolibc.specs)
+ifeq ($(PICOLIBC_SPECS),picolibc.specs)
+# Picolibc is not used.
+else
+CFLAGS += --specs=$(PICOLIBC_SPECS)
+endif
+
 ifeq ($(BOARD),virt)
 ABI=ilp32
 # GCC 10 and lower do not recognise the Zicsr extension in the
@@ -37,9 +48,9 @@ endif
 AFLAGS = -mabi=$(ABI) -march=$(ARCH)
 CFLAGS += -mabi=$(ABI) -march=$(ARCH) -mcmodel=medany
 CFLAGS += -DQEMU_SOC_VIRT
-ifeq ($(PICOLIBC),1)
-CFLAGS += -DUSE_PICOLIBC -D__TMP_FLT_EVAL_METHOD
-endif
```

I still have to work out a couple of details for linking, but this can work as a starting point to graft your makefile changes onto :)

Reply:
I adapted your changes to work on GCC 10 so it now it truly builds out of the box on Ubuntu 22.04 LTS: https://github.com/agatti/micropython/commit/b97731c9c879a53c60acca86e5775044a2c85a93 - can you please check if it builds with the other toolchains you tried earlier?

Reply:
Cool, great progress! Here is the patch I needed to add on top of that commit for it to build and pass tests:

```patch
diff --git i/ports/qemu-riscv/Makefile w/ports/qemu-riscv/Makefile
index 06da71becc..7536150e7f 100644
--- i/ports/qemu-riscv/Makefile
+++ w/ports/qemu-riscv/Makefile
@@ -46,12 +46,14 @@ ARCH=rv32imac
 else
 # Recent GCC versions explicitly require to declare extensions.
 ARCH=rv32imac_zicsr
+ARCH_LD=rv32imac  # binutils may not accept the newer style
 endif
 AFLAGS = -mabi=$(ABI) -march=$(ARCH)
 CFLAGS += $(AFLAGS) -mcmodel=medany
 CFLAGS += -DQEMU_SOC_VIRT
 ARCH_LD ?= $(ARCH)
-LDFLAGS += -mabi=$(ABI) -march-$(ARCH_LD) -T $(LDSCRIPT) -Wl,-EL
+LDSCRIPT = virt.ld
+LDFLAGS += -mabi=$(ABI) -march=$(ARCH_LD) -T $(LDSCRIPT) -Wl,-EL
 SRC_BOARD_O = shared/runtime/gchelper_generic.o setjmp.o
 SRC_BOARD_O += entrypoint.o
 endif
@@ -75,7 +77,6 @@ COPT += -Os -DNDEBUG
 endif
 
 LDFLAGS += $(SPECS_FRAGMENT) -Wl,--gc-sections -Wl,-Map=$(@:.elf=.map)
-LIBS = $(shell $(CC) $(CFLAGS) -print-libgcc-file-name)
 
 SRC_COMMON_C = \
        interrupts.c \
```

From top to bottom:

* `ARCH_LD` - My binutils 2.39 doesn't recognise march=rv32imac_zicsr, have to fall back to the shorter string. Binutils release notes don't mention anything about this changing in newer versions either, so I think it's the right solution for now.
* `LDSCRIPT`/`march-` typo - I'm guessing these crept in when cleaning up the commit, as this shouldn't have worked otherwise. :)
* Removing the `LIBS` line - This does work with it present, but it should be redundant - this is the benefit of linking with CC instead of LD, gcc will pass through what ld needs to find the right system libraries. If it doesn't work then it's probably a sign of a problem somewhere else.

Regarding the rest of the patch, that's a clever idea to check for picolibc.specs. I'm curious what happens if you don't pass this on the command line, though? If the toolchain is already configured for picolibc then I would have thought it would pick this spec automatically.

I'm not sure I see the need for `MICROPY_PICOLIBC` macro. If you include any standard libc header and it's using picolibc then `__PICOLIBC__` will be defined, surely that's the cleanest way to test for it?

Reply:
Thanks for trying that.  I've cleaned things up in https://github.com/agatti/micropython/commit/89a97283e3821873d0a61a6642363a5572cfbad7.

The only thing that won't be carried on is the `FLT_EVAL_METHOD` change in libm.  Besides that, I'd say it's ready for squashing and force pushing once libm is fixed :)

Reply:
> I'm curious what happens if you don't pass this on the command line, though? If the toolchain is already configured for picolibc then I would have thought it would pick this spec automatically.

On Ubuntu 22.04:

```bash
# ...
In file included from ../../py/mpstate.h:29,
                 from ../../py/mpstate.c:27:
/usr/lib/gcc/riscv64-unknown-elf/10.2.0/include/stdint.h:9:16: fatal error: stdint.h: No such file or directory
    9 | # include_next <stdint.h>
      |                ^~~~~~~~~~
compilation terminated.
../../py/malloc.c:27:10: fatal error: stdio.h: No such file or directory
   27 | #include <stdio.h>
      |          ^~~~~~~~~
compilation terminated.
../../py/gc.c:28:10: fatal error: assert.h: No such file or directory
   28 | #include <assert.h>
      |          ^~~~~~~~~~
compilation terminated.
../../py/pystack.c:27:10: fatal error: stdio.h: No such file or directory
   27 | #include <stdio.h>
      |          ^~~~~~~~~
compilation terminated.
# ...
```

The only issue I can see with this is a build failure if both newlib and picolibc are present but picolibc is incorrectly configured.  Figuring out whether a toolchain is configured to use picolibc by default can probably be done via `$(CC) -dumpspecs` and some creative text processing with GNU make functions, but I wonder whether it's really worth it :)